File: src/backend/MasterMind.API/Services/Implementations/FinanceService.cs
using MasterMind.API.Data;
using MasterMind.API.Models.Entities;
using MasterMind.API.Services.Interfaces;
using Microsoft.EntityFrameworkCore;

namespace MasterMind.API.Services.Implementations;

public class FinanceService : IFinanceService
{
    private readonly MasterMindDbContext _context;
    private readonly ILogger<FinanceService> _logger;

    public FinanceService(MasterMindDbContext context, ILogger<FinanceService> logger)
    {
        _context = context;
        _logger = logger;
    }

    public async Task<FinancialSummary> GetFinancialSummaryAsync()
    {
        var currentSession = await _context.Sessions
            .Where(s => s.IsActive)
            .FirstOrDefaultAsync();

        var totalStudents = await _context.Students
            .CountAsync(s => !s.IsDeleted && s.IsActive && (currentSession == null || s.SessionId == currentSession.Id));

        // Calculate total revenue from payments
        var totalRevenue = await _context.Payments
            .Where(p => p.Status == PaymentStatus.Completed)
            .SumAsync(p => (decimal)p.Amount);

        // Calculate pending payments
        var pendingPayments = await _context.StudentFees
            .Where(sf => sf.Status == FeeStatus.Pending || sf.Status == FeeStatus.PartiallyPaid)
            .SumAsync(sf => sf.BalanceAmount);

        // Calculate total expenses (teacher salaries + other expenses)
        var totalSalaries = await _context.TeacherSalaries
            .Where(ts => ts.Status == SalaryStatus.Paid)
            .SumAsync(ts => ts.NetSalary);

        var totalExpenses = totalSalaries;
        var netProfit = totalRevenue - totalExpenses;

        var paidStudents = await _context.StudentFees
            .Where(sf => sf.Status == FeeStatus.Paid)
            .Select(sf => sf.StudentId)
            .Distinct()
            .CountAsync();

        var pendingStudents = await _context.StudentFees
            .Where(sf => sf.Status == FeeStatus.Pending || sf.Status == FeeStatus.PartiallyPaid)
            .Select(sf => sf.StudentId)
            .Distinct()
            .CountAsync();

        var overdueStudents = await _context.StudentFees
            .Where(sf => sf.Status != FeeStatus.Paid && DateOnly.FromDateTime(DateTime.Today) > sf.DueDate)
            .Select(sf => sf.StudentId)
            .Distinct()
            .CountAsync();

        return new FinancialSummary
        {
            TotalRevenue = totalRevenue,
            PendingPayments = pendingPayments,
            Expenses = totalExpenses,
            NetProfit = netProfit,
            TotalStudents = totalStudents,
            PaidStudents = paidStudents,
            PendingStudents = pendingStudents,
            OverdueStudents = overdueStudents
        };
    }

    public async Task<IEnumerable<Payment>> GetRecentPaymentsAsync(int limit = 10)
    {
        return await _context.Payments
            .Include(p => p.StudentFee)
                .ThenInclude(sf => sf.Student)
            .Include(p => p.StudentFee.FeeStructure)
            .Where(p => p.Status == PaymentStatus.Completed)
            .OrderByDescending(p => p.PaymentDate)
            .Take(limit)
            .ToListAsync();
    }

    public async Task<IEnumerable<Payment>> GetPaymentHistoryAsync(int? studentId = null, DateTime? startDate = null, DateTime? endDate = null)
    {
        var query = _context.Payments
            .Include(p => p.StudentFee)
                .ThenInclude(sf => sf.Student)
            .Include(p => p.StudentFee.FeeStructure)
            .AsQueryable();

        if (studentId.HasValue)
        {
            query = query.Where(p => p.StudentFee.StudentId == studentId.Value);
        }

        if (startDate.HasValue)
        {
            query = query.Where(p => p.PaymentDate >= startDate.Value);
        }

        if (endDate.HasValue)
        {
            query = query.Where(p => p.PaymentDate <= endDate.Value);
        }

        return await query
            .OrderByDescending(p => p.PaymentDate)
            .ToListAsync();
    }

    public async Task<Payment> CreatePaymentAsync(CreatePaymentDto paymentDto)
    {
        var studentFee = await _context.StudentFees
            .Include(sf => sf.Student)
            .Include(sf => sf.FeeStructure)
            .FirstOrDefaultAsync(sf => sf.Id == paymentDto.StudentFeeId);

        if (studentFee == null)
        {
            throw new ArgumentException("Student fee not found", nameof(paymentDto.StudentFeeId));
        }

        if (paymentDto.Amount > studentFee.BalanceAmount)
        {
            throw new ArgumentException("Payment amount exceeds balance amount", nameof(paymentDto.Amount));
        }

        var payment = new Payment
        {
            StudentFeeId = paymentDto.StudentFeeId,
            Amount = paymentDto.Amount,
            Method = paymentDto.Method,
            TransactionId = paymentDto.TransactionId,
            ReceiptNumber = paymentDto.InvoiceId,
            Remarks = paymentDto.Description,
            Status = PaymentStatus.Completed
        };

        _context.Payments.Add(payment);

        // Update student fee paid amount and status
        studentFee.PaidAmount += paymentDto.Amount;
        if (studentFee.PaidAmount >= studentFee.FinalAmount)
        {
            studentFee.Status = FeeStatus.Paid;
            studentFee.PaidAmount = studentFee.FinalAmount;
        }
        else
        {
            studentFee.Status = FeeStatus.PartiallyPaid;
        }

        await _context.SaveChangesAsync();
        return payment;
    }

    public async Task<IEnumerable<Payment>> GetPendingPaymentsAsync()
    {
        var pendingStudentFeeIds = await _context.StudentFees
            .Where(sf => sf.Status == FeeStatus.Pending || sf.Status == FeeStatus.PartiallyPaid)
            .Select(sf => sf.Id)
            .ToListAsync();

        return await _context.Payments
            .Include(p => p.StudentFee)
                .ThenInclude(sf => sf.Student)
            .Include(p => p.StudentFee.FeeStructure)
            .Where(p => pendingStudentFeeIds.Contains(p.StudentFeeId))
            .OrderByDescending(p => p.PaymentDate)
            .ToListAsync();
    }

    public async Task<IEnumerable<StudentFee>> GetFeesAsync(int? classId = null, string? status = null, string? month = null)
    {
        var query = _context.StudentFees
            .Include(sf => sf.Student)
                .ThenInclude(s => s.StudentClasses)
                    .ThenInclude(sc => sc.Class)
            .Include(sf => sf.FeeStructure)
            .AsQueryable();

        if (classId.HasValue)
        {
            query = query.Where(sf => sf.Student.StudentClasses.Any(sc => sc.ClassId == classId.Value));
        }

        if (!string.IsNullOrEmpty(status) && Enum.TryParse<FeeStatus>(status, true, out var feeStatus))
        {
            query = query.Where(sf => sf.Status == feeStatus);
        }

        if (!string.IsNullOrEmpty(month))
        {
            query = query.Where(sf => sf.Month == month);
        }

        return await query
            .OrderByDescending(sf => sf.DueDate)
            .ToListAsync();
    }

    public async Task<StudentFee> CreateFeeAsync(CreateFeeDto feeDto)
    {
        var student = await _context.Students
            .Include(s => s.StudentClasses)
                .ThenInclude(sc => sc.Class)
            .FirstOrDefaultAsync(s => s.Id == feeDto.StudentId);

        if (student == null)
        {
            throw new ArgumentException("Student not found", nameof(feeDto.StudentId));
        }

        var feeStructure = await _context.FeeStructures
            .FirstOrDefaultAsync(fs => fs.Id == feeDto.FeeStructureId);

        if (feeStructure == null)
        {
            throw new ArgumentException("Fee structure not found", nameof(feeDto.FeeStructureId));
        }

        var studentFee = new StudentFee
        {
            StudentId = feeDto.StudentId,
            FeeStructureId = feeDto.FeeStructureId,
            Amount = feeStructure.Amount,
            DiscountAmount = feeDto.DiscountAmount,
            DiscountReason = feeDto.DiscountReason,
            FinalAmount = feeStructure.Amount - (feeDto.DiscountAmount ?? 0),
            DueDate = DateOnly.Parse(feeDto.DueDate),
            Status = FeeStatus.Pending,
            Remarks = feeDto.Description,
            Month = feeDto.Month,
            AcademicYear = GetCurrentAcademicYear()
        };

        _context.StudentFees.Add(studentFee);
        await _context.SaveChangesAsync();

        return studentFee;
    }

    public async Task<StudentFee> UpdateFeeAsync(int id, UpdateFeeDto feeDto)
    {
        var studentFee = await _context.StudentFees
            .Include(sf => sf.Student)
                .ThenInclude(s => s.StudentClasses)
                    .ThenInclude(sc => sc.Class)
            .Include(sf => sf.FeeStructure)
            .FirstOrDefaultAsync(sf => sf.Id == id);

        if (studentFee == null)
        {
            throw new ArgumentException("Fee not found", nameof(id));
        }

        // Don't allow updates if fee is already paid
        if (studentFee.Status == FeeStatus.Paid)
        {
            throw new InvalidOperationException("Cannot update a paid fee");
        }

        if (!string.IsNullOrEmpty(feeDto.DueDate))
        {
            studentFee.DueDate = DateOnly.Parse(feeDto.DueDate);
        }

        if (!string.IsNullOrEmpty(feeDto.Description))
        {
            studentFee.Remarks = feeDto.Description;
        }

        if (feeDto.DiscountAmount.HasValue)
        {
            studentFee.DiscountAmount = feeDto.DiscountAmount.Value;
            studentFee.FinalAmount = studentFee.Amount - feeDto.DiscountAmount.Value;
            // BalanceAmount is now computed automatically: FinalAmount - PaidAmount
        }

        await _context.SaveChangesAsync();
        return studentFee;
    }

    public async Task<bool> DeleteFeeAsync(int id)
    {
        var studentFee = await _context.StudentFees
            .Include(sf => sf.Payments)
            .FirstOrDefaultAsync(sf => sf.Id == id);

        if (studentFee == null)
        {
            return false;
        }

        // Don't allow deletion if there are payments
        if (studentFee.Payments.Any())
        {
            throw new InvalidOperationException("Cannot delete fee with existing payments");
        }

        _context.StudentFees.Remove(studentFee);
        await _context.SaveChangesAsync();

        return true;
    }

    public async Task<IEnumerable<StudentFee>> GetOverdueFeesAsync()
    {
        return await _context.StudentFees
            .Include(sf => sf.Student)
                .ThenInclude(s => s.StudentClasses)
                    .ThenInclude(sc => sc.Class)
            .Include(sf => sf.FeeStructure)
            .Where(sf => sf.Status != FeeStatus.Paid && DateOnly.FromDateTime(DateTime.Today) > sf.DueDate)
            .OrderBy(sf => sf.DueDate)
            .ToListAsync();
    }

    public async Task<bool> SendRemindersAsync(List<int>? feeIds = null)
    {
        var query = _context.StudentFees
            .Include(sf => sf.Student)
            .Where(sf => sf.Status != FeeStatus.Paid && DateOnly.FromDateTime(DateTime.Today) > sf.DueDate);

        if (feeIds != null && feeIds.Any())
        {
            query = query.Where(sf => feeIds.Contains(sf.Id));
        }

        var overdueFees = await query.ToListAsync();

        // Here you would implement actual SMS/Email sending logic
        foreach (var fee in overdueFees)
        {
            _logger.LogInformation($"Reminder sent to {fee.Student.ParentName} ({fee.Student.ParentMobile}) for fee {fee.Id}");
        }

        return true;
    }

    public async Task<StudentFee> MarkFeeAsPaidAsync(int id)
    {
        var studentFee = await _context.StudentFees
            .Include(sf => sf.Student)
                .ThenInclude(s => s.StudentClasses)
                    .ThenInclude(sc => sc.Class)
            .Include(sf => sf.FeeStructure)
            .FirstOrDefaultAsync(sf => sf.Id == id);

        if (studentFee == null)
        {
            throw new ArgumentException("Fee not found", nameof(id));
        }

        // Create a payment for the remaining balance
        if (studentFee.BalanceAmount > 0)
        {
            var payment = new Payment
            {
                StudentFeeId = studentFee.Id,
                Amount = studentFee.BalanceAmount,
                Method = PaymentMethod.Cash, // Default method
                Status = PaymentStatus.Completed,
                Remarks = "Marked as paid manually"
            };

            _context.Payments.Add(payment);
        }

        studentFee.Status = FeeStatus.Paid;
        studentFee.PaidAmount = studentFee.FinalAmount;

        await _context.SaveChangesAsync();
        return studentFee;
    }

    public async Task<IEnumerable<FeeStructure>> GetFeeStructuresAsync()
    {
        return await _context.FeeStructures
            .Include(fs => fs.Class)
            .Where(fs => fs.IsActive)
            .OrderBy(fs => fs.Name)
            .ToListAsync();
    }

    public async Task<IEnumerable<ExpenseDto>> GetExpensesAsync(string? category = null, DateTime? startDate = null, DateTime? endDate = null)
    {
        var query = _context.TeacherSalaries
            .Include(ts => ts.Teacher)
            .AsQueryable();

        // Filter by date range if provided
        if (startDate.HasValue)
        {
            query = query.Where(ts => ts.PaymentDate >= startDate.Value);
        }

        if (endDate.HasValue)
        {
            query = query.Where(ts => ts.PaymentDate <= endDate.Value);
        }

        var salaries = await query
            .OrderByDescending(ts => ts.PaymentDate)
            .ToListAsync();

        var expenses = new List<ExpenseDto>();

        foreach (var salary in salaries)
        {
            expenses.Add(new ExpenseDto
            {
                Id = salary.Id,
                Category = "Salary",
                Description = $"Salary for {salary.Teacher.FirstName} {salary.Teacher.LastName} - {salary.Month}",
                Amount = salary.NetSalary,
                PaidTo = $"{salary.Teacher.FirstName} {salary.Teacher.LastName}",
                Date = salary.PaymentDate?.ToString("yyyy-MM-dd") ?? salary.CreatedAt.ToString("yyyy-MM-dd"),
                ReceiptNumber = salary.TransactionId,
                Status = salary.Status.ToString()
            });
        }

        // Filter by category if provided
        if (!string.IsNullOrEmpty(category))
        {
            expenses = expenses.Where(e => e.Category.Equals(category, StringComparison.OrdinalIgnoreCase)).ToList();
        }

        return expenses;
    }

    public async Task<ExpenseDto> CreateExpenseAsync(CreateExpenseRequest expenseDto)
    {
        // For now, we'll create a mock expense since we don't have a dedicated Expense table
        var expense = new ExpenseDto
        {
            Id = DateTime.UtcNow.Ticks,
            Category = expenseDto.Category,
            Description = expenseDto.Description,
            Amount = expenseDto.Amount,
            PaidTo = expenseDto.PaidTo,
            Date = expenseDto.Date,
            ReceiptNumber = expenseDto.ReceiptNumber,
            Status = "Completed"
        };

        _logger.LogInformation($"Expense created: {expenseDto.Category} - {expenseDto.Amount} paid to {expenseDto.PaidTo}");

        return expense;
    }

    public async Task<ExpenseDto> UpdateExpenseAsync(int id, UpdateExpenseRequest expenseDto)
    {
        // For now, we'll simulate the update since we don't have a dedicated Expense table
        _logger.LogInformation($"Expense {id} updated: {expenseDto.Category} - {expenseDto.Amount}");

        var expense = new ExpenseDto
        {
            Id = id,
            Category = expenseDto.Category,
            Description = expenseDto.Description,
            Amount = expenseDto.Amount,
            PaidTo = expenseDto.PaidTo,
            Date = expenseDto.Date,
            ReceiptNumber = expenseDto.ReceiptNumber,
            Status = "Completed"
        };

        return expense;
    }

    public async Task<bool> DeleteExpenseAsync(int id)
    {
        // For now, we'll simulate the deletion since we don't have a dedicated Expense table
        _logger.LogInformation($"Expense {id} deleted");
        return true;
    }

    public async Task<IEnumerable<string>> GetExpenseCategoriesAsync()
    {
        return new List<string>
        {
            "Salary",
            "Rent",
            "Utilities",
            "Supplies",
            "Maintenance",
            "Marketing",
            "Other"
        };
    }

    public async Task<IEnumerable<ExpenseSummaryDto>> GetExpenseSummaryAsync(DateTime? startDate = null, DateTime? endDate = null)
    {
        var query = _context.TeacherSalaries.AsQueryable();

        // Filter by date range if provided
        if (startDate.HasValue)
        {
            query = query.Where(ts => ts.PaymentDate >= startDate.Value);
        }

        if (endDate.HasValue)
        {
            query = query.Where(ts => ts.PaymentDate <= endDate.Value);
        }

        var salaries = await query
            .Where(ts => ts.Status == SalaryStatus.Paid)
            .ToListAsync();

        var summary = new List<ExpenseSummaryDto>
        {
            new ExpenseSummaryDto
            {
                Category = "Salary",
                TotalAmount = salaries.Sum(s => s.NetSalary),
                Count = salaries.Count
            }
        };

        // Add other categories with zero amounts for now
        summary.Add(new ExpenseSummaryDto { Category = "Rent", TotalAmount = 0, Count = 0 });
        summary.Add(new ExpenseSummaryDto { Category = "Utilities", TotalAmount = 0, Count = 0 });
        summary.Add(new ExpenseSummaryDto { Category = "Supplies", TotalAmount = 0, Count = 0 });
        summary.Add(new ExpenseSummaryDto { Category = "Maintenance", TotalAmount = 0, Count = 0 });
        summary.Add(new ExpenseSummaryDto { Category = "Marketing", TotalAmount = 0, Count = 0 });
        summary.Add(new ExpenseSummaryDto { Category = "Other", TotalAmount = 0, Count = 0 });

        return summary;
    }

    public async Task<FinancialReport> GenerateReportAsync(DateTime startDate, DateTime endDate)
    {
        var payments = await _context.Payments
            .Include(p => p.StudentFee)
                .ThenInclude(sf => sf.Student)
            .Where(p => p.Status == PaymentStatus.Completed &&
                       p.PaymentDate >= startDate &&
                       p.PaymentDate <= endDate)
            .ToListAsync();

        var totalRevenue = payments.Sum(p => p.Amount);

        var expenses = await _context.TeacherSalaries
            .Where(ts => ts.Status == SalaryStatus.Paid &&
                       ts.PaymentDate >= startDate &&
                       ts.PaymentDate <= endDate)
            .ToListAsync();

        var totalExpenses = expenses.Sum(e => e.NetSalary);

        var report = new FinancialReport
        {
            ReportId = DateTime.UtcNow.Ticks,
            GeneratedAt = DateTime.UtcNow,
            Period = new ReportPeriod
            {
                StartDate = startDate.ToString("yyyy-MM-dd"),
                EndDate = endDate.ToString("yyyy-MM-dd")
            },
            Data = new ReportData
            {
                TotalRevenue = totalRevenue,
                TotalExpenses = totalExpenses,
                NetProfit = totalRevenue - totalExpenses,
                Payments = payments.Select(p => new PaymentDto
                {
                    Id = p.Id,
                    StudentId = p.StudentFee.StudentId,
                    StudentName = $"{p.StudentFee.Student.FirstName} {p.StudentFee.Student.LastName}",
                    Amount = p.Amount,
                    Date = p.PaymentDate.ToString("yyyy-MM-dd"),
                    Status = p.Status.ToString(),
                    Method = p.Method.ToString(),
                    Description = p.StudentFee.FeeStructure.Name,
                    InvoiceId = p.ReceiptNumber
                }).ToList(),
                Expenses = expenses.Select(e => new ExpenseDto
                {
                    Id = e.Id,
                    Category = "Salary",
                    Description = $"Salary for {e.Teacher.FirstName} {e.Teacher.LastName}",
                    Amount = e.NetSalary,
                    PaidTo = $"{e.Teacher.FirstName} {e.Teacher.LastName}",
                    Date = e.PaymentDate?.ToString("yyyy-MM-dd") ?? "",
                    ReceiptNumber = e.TransactionId,
                    Status = e.Status.ToString()
                }).ToList()
            }
        };

        return report;
    }

    private string GetCurrentAcademicYear()
    {
        var currentYear = DateTime.Now.Year;
        var nextYear = currentYear + 1;
        return $"{currentYear}-{nextYear}";
    }
}

File: src/frontend/mastermind-web/src/services/financeService.ts
import { apiService } from './apiService'

// Environment variable to control mock API usage
const USE_MOCK_API = import.meta.env.VITE_USE_MOCK_API === 'true'

// Interfaces
export interface FinancialSummary {
  totalRevenue: number
  pendingPayments: number
  expenses: number
  netProfit: number
  totalStudents: number
  paidStudents: number
  pendingStudents: number
  overdueStudents: number
}

export interface Payment {
  id: number
  studentId: number
  studentName: string
  amount: number
  paymentMethod: string
  date: string
  status: 'Completed' | 'Pending' | 'Failed'
  description?: string
}

export interface Fee {
  id: number
  studentId: number
  studentName: string
  classId: number
  className: string
  feeType: string
  amount: number
  dueDate: string
  status: 'Paid' | 'Pending' | 'Overdue'
  description?: string
}

export interface Expense {
  id: number
  category: string
  description: string
  amount: number
  paidTo: string
  date: string
  receiptNumber?: string
}

export interface CreateExpenseDto {
  category: string
  description: string
  amount: number
  paidTo: string
  date: string
  receiptNumber?: string
}

export const financeService = {
  // ===== FINANCIAL SUMMARY METHODS =====

  // Get financial summary
  async getFinancialSummary(): Promise<FinancialSummary> {
    if (USE_MOCK_API) {
      await new Promise(resolve => setTimeout(resolve, 500))
      return {
        totalRevenue: 245000,
        pendingPayments: 45000,
        expenses: 85000,
        netProfit: 160000,
        totalStudents: 150,
        paidStudents: 120,
        pendingStudents: 25,
        overdueStudents: 5
      }
    }

    const response = await apiService.get('/finance/summary')
    return response.data
  },

  // ===== PAYMENT METHODS =====

  // Get recent payments
  async getRecentPayments(limit: number = 10): Promise<Payment[]> {
    if (USE_MOCK_API) {
      await new Promise(resolve => setTimeout(resolve, 500))
      return [
        {
          id: 1,
          studentId: 1,
          studentName: 'John Doe',
          amount: 5000,
          paymentMethod: 'Cash',
          date: '2024-01-15',
          status: 'Completed',
          description: 'Monthly tuition fee'
        },
        {
          id: 2,
          studentId: 2,
          studentName: 'Jane Smith',
          amount: 4500,
          paymentMethod: 'Online',
          date: '2024-01-14',
          status: 'Completed',
          description: 'Monthly tuition fee'
        }
      ]
    }

    const response = await apiService.get(`/finance/payments?limit=${limit}`)
    return response.data
  },

  // Get pending payments
  async getPendingPayments(): Promise<Payment[]> {
    if (USE_MOCK_API) {
      await new Promise(resolve => setTimeout(resolve, 500))
      return [
        {
          id: 3,
          studentId: 3,
          studentName: 'Bob Johnson',
          amount: 4000,
          paymentMethod: 'Pending',
          date: '2024-01-20',
          status: 'Pending',
          description: 'Monthly tuition fee'
        }
      ]
    }

    const response = await apiService.get('/finance/payments/pending')
    return response.data
  },

  // ===== FEE METHODS =====

  // Create a new fee with enhanced fee type support
  async createFee(data: any): Promise<any> {
    if (USE_MOCK_API) {
      await new Promise(resolve => setTimeout(resolve, 1000))
      return {
        success: true,
        message: 'Fee created successfully',
        data: {
          id: Date.now(),
          studentId: data.studentId,
          feeCategory: data.feeCategory,
          amount: data.amount,
          status: 'Created'
        }
      }
    }

    const response = await apiService.post('/finance/fees', data)
    return response.data
  },

  // Get all fees
  async getFees(): Promise<Fee[]> {
    if (USE_MOCK_API) {
      await new Promise(resolve => setTimeout(resolve, 500))
      return []
    }

    const response = await apiService.get('/finance/fees')
    return response.data || []
  },

  // Get overdue fees
  async getOverdueFees(): Promise<Fee[]> {
    if (USE_MOCK_API) {
      await new Promise(resolve => setTimeout(resolve, 500))
      return [
        {
          id: 2,
          studentId: 2,
          studentName: 'Jane Smith',
          classId: 2,
          className: 'Class 9',
          feeType: 'Tuition Fee',
          amount: 4500,
          dueDate: '2024-01-10',
          status: 'Overdue',
          description: 'Monthly tuition fee'
        }
      ]
    }

    const response = await apiService.get('/finance/fees/overdue')
    return response.data || []
  },

  // Send reminders for overdue fees
  async sendReminders(feeIds?: number[]): Promise<boolean> {
    if (USE_MOCK_API) {
      await new Promise(resolve => setTimeout(resolve, 1000))
      return true
    }

    const response = await apiService.post('/finance/fees/reminders', { feeIds })
    return response.data
  },

  // Mark fee as paid
  async markFeeAsPaid(feeId: number): Promise<Fee> {
    if (USE_MOCK_API) {
      await new Promise(resolve => setTimeout(resolve, 500))
      return {
        id: feeId,
        studentId: 2,
        studentName: 'Jane Smith',
        classId: 2,
        className: 'Class 9',
        feeType: 'Tuition Fee',
        amount: 4500,
        dueDate: '2024-01-10',
        status: 'Paid',
        description: 'Monthly tuition fee'
      }
    }

    const response = await apiService.post(`/finance/fees/${feeId}/mark-paid`)
    return response.data
  },

  // ===== EXPENSE METHODS =====

  // Get expenses
  async getExpenses(filters?: {
    category?: string
    startDate?: string
    endDate?: string
  }): Promise<Expense[]> {
    if (USE_MOCK_API) {
      await new Promise(resolve => setTimeout(resolve, 500))
      return [
        {
          id: 1,
          category: 'Salary',
          description: 'Teacher salaries',
          amount: 50000,
          paidTo: 'Various Teachers',
          date: '2024-01-01',
          receiptNumber: 'REC-001'
        },
        {
          id: 2,
          category: 'Rent',
          description: 'Monthly rent',
          amount: 20000,
          paidTo: 'Landlord',
          date: '2024-01-05',
          receiptNumber: 'REC-002'
        }
      ]
    }

    const params = new URLSearchParams()
    if (filters?.category) params.append('category', filters.category)
    if (filters?.startDate) params.append('startDate', filters.startDate)
    if (filters?.endDate) params.append('endDate', filters.endDate)

    const response = await apiService.get(`/finance/expenses?${params.toString()}`)
    return response.data || []
  },

  // Create expense
  async createExpense(data: CreateExpenseDto): Promise<Expense> {
    if (USE_MOCK_API) {
      await new Promise(resolve => setTimeout(resolve, 500))
      return {
        id: Date.now(),
        category: data.category,
        description: data.description,
        amount: data.amount,
        paidTo: data.paidTo,
        date: data.date,
        receiptNumber: data.receiptNumber
      }
    }

    const response = await apiService.post('/finance/expenses', data)
    return response.data
  },

  // ===== REPORT METHODS =====

  // Generate financial report
  async generateReport(startDate: string, endDate: string): Promise<any> {
    if (USE_MOCK_API) {
      await new Promise(resolve => setTimeout(resolve, 1000))
      return {
        reportId: Date.now(),
        generatedAt: new Date().toISOString(),
        period: { startDate, endDate },
        data: {
          totalRevenue: 245000,
          totalExpenses: 85000,
          netProfit: 160000,
          payments: [],
          expenses: [],
          fees: [],
          summary: {
            totalStudents: 150,
            paidStudents: 120,
            pendingStudents: 25,
            overdueStudents: 5
          }
        }
      }
    }

    const response = await apiService.post('/finance/reports/generate', {
      startDate,
      endDate
    })
    return response.data
  },

  // Get reports
  async getReports(): Promise<any[]> {
    if (USE_MOCK_API) {
      await new Promise(resolve => setTimeout(resolve, 500))
      return [
        {
          id: 1,
          type: 'Monthly Report',
          period: 'January 2024',
          generatedAt: new Date().toISOString(),
          data: {}
        }
      ]
    }

    const response = await apiService.get('/finance/reports')
    return response.data || []
  },

  // ===== FEE COLLECTION METHODS =====

  // Setup student fee (Monthly or Full Course)
  async setupStudentFee(data: any): Promise<any> {
    if (USE_MOCK_API) {
      await new Promise(resolve => setTimeout(resolve, 1000))
      return {
        studentId: data.studentId,
        studentName: 'John Doe',
        feeType: data.feeType,
        totalAmount: data.totalAmount,
        monthlyAmount: data.monthlyAmount,
        numberOfInstallments: data.numberOfInstallments,
        startDate: data.startDate,
        endDate: data.endDate,
        status: 'Active'
      }
    }

    const response = await apiService.post('/finance/fees/setup', data)
    return response.data
  },

  // Collect payment
  async collectPayment(data: any): Promise<any> {
    if (USE_MOCK_API) {
      await new Promise(resolve => setTimeout(resolve, 500))
      return {
        id: Date.now(),
        receiptNumber: `RCP${Date.now()}`,
        studentName: 'John Doe',
        studentClass: 'Class 10',
        totalAmount: data.amount,
        paidAmount: data.amount,
        balanceAmount: 0,
        paymentMethod: data.paymentMethod,
        receiptDate: new Date().toISOString(),
        paymentStatus: 'Completed',
        collectedBy: 'Admin',
        remarks: data.remarks || '',
        receiptItems: [
          {
            feeType: data.feeType,
            amount: data.amount,
            dueDate: data.dueDate,
            status: 'Paid'
          }
        ]
      }
    }

    const response = await apiService.post('/finance/payments/collect', data)
    return response.data
  },

  // Get receipt by ID
  async getReceipt(receiptId: number): Promise<any> {
    if (USE_MOCK_API) {
      await new Promise(resolve => setTimeout(resolve, 500))
      return {
        id: receiptId,
        receiptNumber: `RCP${receiptId}`,
        studentName: 'John Doe',
        studentClass: 'Class 10',
        totalAmount: 5000,
        paidAmount: 5000,
        balanceAmount: 0,
        paymentMethod: 'Cash',
        receiptDate: new Date().toISOString(),
        paymentStatus: 'Completed',
        collectedBy: 'Admin',
        remarks: 'Payment completed successfully',
        receiptItems: []
      }
    }

    const response = await apiService.get(`/finance/payments/receipt/${receiptId}`)
    return response.data
  },

  // Send receipt via email
  async sendReceiptEmail(receiptId: number, email: string): Promise<boolean> {
    if (USE_MOCK_API) {
      await new Promise(resolve => setTimeout(resolve, 1000))
      return true
    }

    const response = await apiService.post(`/finance/payments/receipt/${receiptId}/send-email`, { email })
    return response.data
  },

  // Get student fee details
  async getStudentFeeDetails(studentId: number): Promise<any> {
    if (USE_MOCK_API) {
      await new Promise(resolve => setTimeout(resolve, 500))
      return {
        studentId: studentId,
        studentName: 'John Doe',
        studentClass: 'Class 10',
        parentName: 'Parent Name',
        parentEmail: 'parent@example.com',
        parentMobile: '+91 9876543210',
        pendingFees: [
          {
            studentFeeId: 1,
            feeType: 'Tuition Fee',
            amount: 5000,
            dueDate: '2024-02-01',
            status: 'Pending',
            isOverdue: false
          }
        ]
      }
    }

    const response = await apiService.get(`/finance/students/${studentId}/fees`)
    return response.data
  },

  // Get fee structures
  async getFeeStructures(): Promise<any[]> {
    if (USE_MOCK_API) {
      await new Promise(resolve => setTimeout(resolve, 500))
      return [
        {
          id: 1,
          name: 'Class 10 - Monthly Tuition',
          type: 'Monthly',
          category: 'Tuition',
          amount: 5000,
          frequency: 'Monthly',
          className: 'Class 10',
          description: 'Monthly tuition fee for Class 10',
          academicYear: '2024-25'
        }
      ]
    }

    const response = await apiService.get('/fees/structures')
    return response.data
  }
}
